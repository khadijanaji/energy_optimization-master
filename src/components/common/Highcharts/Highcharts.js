// @flow

import * as React from "react";
import * as highcharts from "highcharts";
import DarkUnicaTheme from "./themes/DarkUnica";

import "./style.scss";

export type HighChartsPropsType = {
    container: string,
    options: any,
    modules?: any,
    domProps?: any,
    theme?: string,
    status: string
}

const updateChartOptions = (currentChart: any, newChartOptions: any) => {
  const [xAxis] = currentChart.axes;
  xAxis.update({"categories": newChartOptions.xAxis.categories});
  currentChart.update({"tooltip": newChartOptions.tooltip});
  if (newChartOptions.xAxis.isCustomYearly) {
    xAxis.update({"tickPositions": newChartOptions.xAxis.tickPositions});
  } else {
    xAxis.update({"tickPositions": undefined});
  }
  xAxis.isDirty = true;
};

class Highcharts extends React.PureComponent<HighChartsPropsType> {
  componentDidMount () {
    const {container, modules, options, theme, status} = this.props;
    if (modules) {
      modules.forEach((module) => {
        module(Highcharts);
      });
    }
    highcharts.setOptions(theme || DarkUnicaTheme);
    const newOptions = {
      ...options,
      "loading": {
        "hideDuration": 1000,
        "showDuration": 1000,
        "style": {"backgroundColor": "#263237"}
      },
      "credits": {"enabled": false},
      "yAxis": {
        "labels": {"format": "{value}"},
        "title": ""
      }
    };
    this.chart = new highcharts.chart(
      container,
      newOptions
    );
    if (status === "Loading") {
      this.chart.showLoading();
    }
  }

  componentWillReceiveProps ({status, options}: HighChartsPropsType) {
    if (status === "Loading") {
      this.chart.showLoading();
    } else {
      const {series = []} = options;
      series.forEach(({data}, ind) => {
        this.chart.series[ind].setData(data);
      });
      this.chart.hideLoading();
      updateChartOptions(this.chart, options);
      this.chart.redraw();
    }
  }

  componentWillUnmount () {
    this.chart.destroy();
  }

  chart: any

  render () {
    const {container, domProps} = this.props;
    return (<div
      id={container}
      {...domProps}
    />);
  }
}
export default Highcharts;
