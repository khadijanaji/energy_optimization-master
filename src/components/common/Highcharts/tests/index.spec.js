/* globals describe it expect */

import {shallow} from "enzyme";
import React from "react";
import {setOptions, chart} from "highcharts";

import Highcharts from "../";

const container = "container-id";
const options = {
  "series": [
    {"data": [1]},
    {"data": [1]}
  ],
  "xAxis": [{"categories": []}],
  "axes": []
};
const newOptions = {
  "series": [
    {
      "data": [
        1,
        1
      ]
    },
    {
      "data": [
        1,
        1
      ]
    }
  ],
  "xAxis": [
    {
      "categories": [],
      "options": {"tickInterval": 5}
    }
  ]
};
const domProps = {"className": "testClassName"};
const modules = [(element) => element];
const theme = {};

jest.mock("highcharts", () => ({
  "setOptions": jest.fn(),
  "chart": jest.fn(() => ({
    "showLoading": jest.fn(),
    "hideLoading": jest.fn(),
    "series": [
      {"setData": jest.fn()},
      {"setData": jest.fn()}
    ],
    "xAxis": [
      {
        "categories": [],
        "options": {"tickInterval": 5}
      }
    ],
    "update": jest.fn(),
    "axes": [{"update": jest.fn()}],
    "redraw": jest.fn()
  }))
}));

jest.mock("../themes/DarkUnica", () => ({}));

describe("Component", () => {
  describe("Highcharts Render", () => {
    it("With all props should render correctly", () => {
      const wrapper = shallow(<Highcharts
        container={container}
        domProps={domProps}
        modules={modules}
        options={options}
        theme={theme}
      />);
      expect(wrapper).toMatchSnapshot();
    });

    it("Verify that highcharts functions are called", () => {
      const wrapper = shallow(<Highcharts
        container={container}
        domProps={domProps}
        modules={modules}
        options={options}
        theme={theme}
      />);

      const destroy = jest.fn();
      wrapper.instance().chart = {destroy};
      wrapper.instance().componentWillUnmount();

      expect(setOptions).toHaveBeenCalled();
      expect(setOptions.mock.calls[0]).toEqual([theme]);
      expect(chart).toHaveBeenCalled();
      expect(chart.mock.calls[0]).toMatchSnapshot();

      expect(destroy).toHaveBeenCalled();
      expect(destroy.mock.calls[0]).toEqual([]);
    });
    it("With Loading status, Should call showLoading", () => {
      const wrapper = shallow(<Highcharts
        container={container}
        domProps={domProps}
        modules={modules}
        options={options}
        status={"Loading"}
        theme={theme}
      />);

      const {showLoading} = wrapper.instance().chart;

      expect(showLoading).toHaveBeenCalled();
      expect(showLoading.mock.calls[0]).toEqual([]);
    });
    it("When props changed and status is Success: Should update data", () => {
      const wrapper = shallow(<Highcharts
        container={container}
        domProps={domProps}
        modules={modules}
        options={options}
        status={"Loading"}
        theme={theme}
      />);


      const {hideLoading, series} = wrapper.instance().chart;

      wrapper.instance().componentWillReceiveProps({
        "status": "Success",
        "options": newOptions
      });

      expect(hideLoading).toHaveBeenCalled();
      expect(hideLoading.mock.calls[0]).toEqual([]);

      expect(series[0].setData).toHaveBeenCalled();
      expect(series[0].setData.mock.calls[0]).toEqual([newOptions.series[0].data]);

      expect(series[1].setData).toHaveBeenCalled();
      expect(series[1].setData.mock.calls[0]).toEqual([newOptions.series[1].data]);
    });
    it("When props changed and status still Loading: Should show loading", () => {
      const wrapper = shallow(<Highcharts
        container={container}
        domProps={domProps}
        modules={modules}
        options={options}
        status={"Loading"}
        theme={theme}
      />);


      const {showLoading} = wrapper.instance().chart;

      wrapper.instance().componentWillReceiveProps({
        "status": "Loading",
        "options": {}
      });

      expect(showLoading).toHaveBeenCalled();
      expect(showLoading.mock.calls[0]).toEqual([]);
    });
  });
});
