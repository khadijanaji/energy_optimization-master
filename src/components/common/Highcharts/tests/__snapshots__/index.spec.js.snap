// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Highcharts Render Verify that highcharts functions are called 1`] = `
Array [
  "container-id",
  Object {
    "axes": Array [],
    "credits": Object {
      "enabled": false,
    },
    "loading": Object {
      "hideDuration": 1000,
      "showDuration": 1000,
      "style": Object {
        "backgroundColor": "#263237",
      },
    },
    "series": Array [
      Object {
        "data": Array [
          1,
        ],
      },
      Object {
        "data": Array [
          1,
        ],
      },
    ],
    "xAxis": Array [
      Object {
        "categories": Array [],
      },
    ],
    "yAxis": Object {
      "labels": Object {
        "format": "{value}",
      },
      "title": "",
    },
  },
]
`;

exports[`Component Highcharts Render With all props should render correctly 1`] = `
<div
  className="testClassName"
  id="container-id"
/>
`;
